### PR TITLE
Improve sample types

### PR DESCRIPTION
**Pull request by doasync**
fix target unit replacement,
fix clock without source,
fix IfAssignable helper,
fix type inference of events returned from map/filter/merge

### CODE REVIEW

File: src/types/__tests__/effector/sample/sample.test.ts [hunk 1]
```diff
@@ -158,7 +300,14 @@ test('generic edge cases', () => {
   }
   expect(typecheck).toMatchInlineSnapshot(`
     "
-    no errors
+    Type 'Store<[any]> | Store<[any, ...any[]]> | Event<any>' is not assignable to type 'Store<A>'.
```
Comment (zerobias) [inline src/types/__tests__/effector/sample/sample.test.ts:303]: regression

**Review by zerobias — CHANGES_REQUESTED** [review]
we need types for guard more than another pr for sample, especially the one with new regression

File: src/types/__tests__/effector/sample/sample.test.ts (hunk 1 above)
Comment (doasync) [reply src/types/__tests__/effector/sample/sample.test.ts:303]: Yep, I'm aware of this. The error appeared again after fixing the types for clock without source. I've tried to fix it, but so far without success.

Comment (zerobias) [reply src/types/__tests__/effector/sample/sample.test.ts:303]: That's a task for further releases, not for `21.8.0`

Comment (doasync) [reply src/types/__tests__/effector/sample/sample.test.ts]: done

**Comment by doasync** [comment]
The regression is fixed. Here are types for guard: https://github.com/effector/effector/pull/447

**Comment by doasync** [comment]
I've removed explicit generics validation from this PR. It's too complicated and can cause unexpected errors.

File: packages/effector/index.d.ts [hunk 2]
```diff
@@ -730,275 +720,225 @@ type GetResultSCF<S, C, F extends AnyFn> = S extends Store<any> | Combinable
...
   fn: F,
-  target: MultiTarget<ReturnType<F>, T, X>
-  name?: string
+  target: MultiTarget<T, ReturnType<F>>,
+  name?: string,
   greedy?: boolean
 }): T
 
-// Last overload
+// sample's last overload (for readable error messages)
```
Comment (zerobias) [inline packages/effector/index.d.ts:914]: 💯

**Review by zerobias — APPROVED** [review]
Great work, thanks! 👍
